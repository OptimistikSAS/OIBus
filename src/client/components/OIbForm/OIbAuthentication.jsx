import React from 'react'
import PropTypes from 'prop-types'
import { Row, Col } from 'reactstrap'

import OIbText from './OIbText.jsx'
import OIbPassword from './OIbPassword.jsx'
import OIbSelect from './OIbSelect.jsx'
import OIbTitle from './OIbTitle.jsx'
import { notEmpty, hasLengthBetween } from '../../../services/validation.service'

/*
  OIBAuthentication is a form reused in several places. Can manage user/password (default)
  or accessKey/secretKey as well as authentication type.
*/
const OIbAuthentication = ({ value, name, onChange, mode, label }) => {
  const handleChange = (attributeName, newValue, valid) => {
    onChange(`${name}.${attributeName}`, newValue, valid)
  }
  const validation = {
    accessKey: notEmpty(),
    secretKey: hasLengthBetween(0, 256),
    username: notEmpty(),
    password: hasLengthBetween(0, 256),
  }
  return [
    <OIbTitle label={label} key="title">
      <div>
        <p>Authentication parameters</p>
        <p>Please fill the user and password to connect this application</p>
      </div>
    </OIbTitle>,
    mode === 'user' ? (
      [
        <Row key="type">
          <Col md="2">
            <OIbSelect
              label="Type"
              onChange={handleChange}
              value={value.type}
              options={['Basic']}
              defaultValue="Basic"
              name="type"
            />
          </Col>
        </Row>,
        <Row key="user">
          <Col md="4">
            <OIbText
              label="User name"
              onChange={handleChange}
              value={value.username}
              valid={validation.username}
              name="username"
            />
          </Col>
          <Col md="4">
            <OIbPassword
              label="Password"
              onChange={handleChange}
              value={value.password}
              valid={validation.password}
              name="password"
            />
          </Col>
        </Row>,
      ]
    ) : (
      <Row key="accessKey">
        <Col md="4">
          <OIbText
            label="Access Key"
            onChange={handleChange}
            value={value.accessKey}
            valid={validation.accessKey}
            name="accessKey"
          />
        </Col>
        <Col md="4">
          <OIbPassword
            label="Secret Key"
            onChange={handleChange}
            value={value.secretKey}
            valid={validation.secretKey}
            name="secretKey"
          />
        </Col>
      </Row>
    ),
  ]
}

OIbAuthentication.propTypes = {
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.object,
  mode: PropTypes.oneOf(['accessKey', 'user']),
  label: PropTypes.string,
}

OIbAuthentication.defaultProps = {
  value: { type: 'basic', username: '', password: '', accessKey: '', secretKey: '' },
  mode: 'user',
  label: 'Authentication',
}

export default OIbAuthentication
