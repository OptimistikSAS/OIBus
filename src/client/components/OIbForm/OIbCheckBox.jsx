import React from 'react'
import PropTypes from 'prop-types'
import { FormGroup, Label, Input, CustomInput } from 'reactstrap'

const OIbCheckBox = ({ label, value, name, onChange, defaultValue, switchButton }) => {
  // if no value was found, load the context with the default value
  // (this will cause a rerender with the correct value)
  React.useLayoutEffect(() => {
    if (value === null) onChange(name, defaultValue)
  }, [value])
  const handleChange = (event) => {
    const { target } = event
    const { checked } = target
    onChange(name, checked, null)
  }
  if (value === null) return null
  return (
    <FormGroup>
      {switchButton ? (
        <CustomInput
          type="switch"
          id={name}
          name={name}
          label={label}
          onChange={handleChange}
          checked={value}
          color="secondary"
        />
      ) : (
        <>
          <Label>{label}</Label>
          <Input
            className="oi-form-input"
            type="checkbox"
            id={name}
            name={name}
            onChange={handleChange}
            checked={value}
            style={{ position: 'relative', left: '1.5rem' }}
          />
        </>
      )}
    </FormGroup>
  )
}
OIbCheckBox.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.bool,
  defaultValue: PropTypes.bool.isRequired,
  switchButton: PropTypes.bool,
}

OIbCheckBox.defaultProps = { value: null, switchButton: false }

export default OIbCheckBox
