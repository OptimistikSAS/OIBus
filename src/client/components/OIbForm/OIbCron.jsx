import React from 'react'
import PropTypes from 'prop-types'
import { FormGroup, FormFeedback, FormText, Label, Input } from 'reactstrap'

const OIbCron = ({ label, help, valid, value, name, onChange, defaultValue }) => {
  const types = ['year', 'month', 'day', 'hour', 'minute', 'sec', 'msec']
  React.useEffect(() => {
    if (value === null) onChange(name, defaultValue)
  }, [value])

  // insert part change to value at index
  // or at end of string
  const newCronValue = (newVal, index) => {
    const parts = value.split(' ')
    if (parts.length === 0) {
      return newVal
    }
    if (parts.length < index) {
      parts.push(newVal)
      return parts.filter((e) => e !== '').join(' ')
    }
    parts[index] = newVal
    return parts.filter((e) => e !== '').join(' ')
  }

  const handleChange = (event) => {
    const { target } = event
    const { value: newVal, name: typeName } = target
    const index = types.indexOf(typeName)

    const newFinalValue = newCronValue(newVal, index)

    onChange(name, newFinalValue, valid(newFinalValue))
  }
  // if no label, we are in a table so we need to minimize the row height
  const style = label ? null : { marginBottom: 0 }

  const validCheck = valid(value)
  // if value is null, no need to render
  if (value === null) return null

  // create capitalized placeholder
  const capitalize = (s) => s.charAt(0).toUpperCase() + s.slice(1)

  // get part based on index
  const valuePart = (v, i) => {
    if (v.split(' ').length > i) {
      return v.split(' ')[i]
    }
    return ''
  }

  // render input part based on index
  const renderInput = (index) => (
    <Input
      key={index}
      className="oi-form-input oi-cron-input"
      type="text"
      id={`${name}.${types[index]}`}
      name={types[index]}
      invalid={validCheck !== null}
      onChange={handleChange}
      value={valuePart(value, index)}
      placeholder={capitalize(types[index])}
    />
  )

  return (
    <FormGroup style={style}>
      {label && <Label for={name}>{label}</Label>}
      {types.map((_, index) => renderInput(index))}
      <FormFeedback>{validCheck}</FormFeedback>
      {help && <FormText>{help}</FormText>}
    </FormGroup>
  )
}
OIbCron.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  help: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  onChange: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  defaultValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  valid: PropTypes.func,
}
OIbCron.defaultProps = {
  valid: () => null,
  label: null,
  help: null,
  value: null,
  defaultValue: '',
}

export default OIbCron
