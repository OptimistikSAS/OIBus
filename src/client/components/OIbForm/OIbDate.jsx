import React from 'react'
import PropTypes from 'prop-types'
import { DateTime } from 'luxon'
import { FormGroup, FormFeedback, FormText, Label, Input } from 'reactstrap'

const OIbDate = ({ label, help, valid, value, name, onChange, inline, hidden, maxDateString }) => {
  const handleChange = (event) => {
    const { target } = event
    const { value: newVal } = target
    onChange(name, newVal, valid(newVal))
  }

  const style = label ? null : { marginBottom: 0 }
  if (inline && style) {
    style.display = 'inline-block'
  }
  if (hidden) return null
  const validCheck = valid(value)
  const formattedDate = DateTime.fromISO(new Date(value).toISOString()).toFormat('yyyy-MM-dd HH:mm')
  return (
    <FormGroup style={style}>
      {label && <Label for={name}>{label}</Label>}
      <Input
        className="oi-form-input"
        type="datetime-local"
        id={name}
        max={maxDateString}
        value={formattedDate}
        placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
        required
        onChange={handleChange}
      />
      <FormFeedback>{validCheck}</FormFeedback>
      {help && <FormText>{help}</FormText>}
    </FormGroup>
  )
}
OIbDate.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  help: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  onChange: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.object]),
  valid: PropTypes.func,
  inline: PropTypes.bool,
  hidden: PropTypes.bool,
  maxDateString: PropTypes.string,
}
OIbDate.defaultProps = {
  valid: () => null,
  label: null,
  help: null,
  value: null,
  inline: false,
  hidden: false,
  maxDateString: null,
}

export default OIbDate
