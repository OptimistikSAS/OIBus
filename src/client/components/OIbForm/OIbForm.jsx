import React from 'react'
import PropTypes from 'prop-types'
import { Row, Col } from 'reactstrap'
import humanizeString from 'humanize-string'

import * as Controls from './index'
import OIbTable from './OIbTable.jsx'

const OIbForm = ({ schema, onChange, values, name: configName }) => {
  const { form } = schema
  /* transform into arrays of line with array of col */
  const rows = []
  let rowNum = -1
  Object.entries(form).forEach(([controlName, parameters]) => {
    const { newRow = true, ...rest } = parameters
    if (newRow || rowNum === 0) {
      rows.push([])
      rowNum += 1
    }
    if (!rest.label) rest.label = humanizeString(controlName)
    rows[rowNum].push({ name: controlName, ...rest })
  })
  return rows.map((cols) => (
    <Row key={cols[0].name}>
      {cols.map((col) => {
        const { type, name, ...rest } = col
        if (!rest.md) rest.md = (['OItype', 'OIbAuthentication'].includes(type)) ? 12 : 4
        const Control = (type === 'OIbTable') ? OIbTable : Controls[type]
        rest.value = values[name]
        return (
          <Col md={rest.md} key={name}>
            {/* eslint-disable-next-line react/jsx-props-no-spreading */}
            <Control onChange={onChange} name={`${configName}.${name}`} {...rest} />
          </Col>
        )
      })}
    </Row>
  ))
}

OIbForm.propTypes = {
  schema: PropTypes.object.isRequired,
  values: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
}

export default OIbForm
