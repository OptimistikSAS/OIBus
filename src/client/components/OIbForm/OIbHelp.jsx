import React from 'react'
import PropTypes from 'prop-types'
import { UncontrolledCollapse, Button, Row } from 'reactstrap'

const Help = ({ children }) => (
  <Row>
    <Button color="link" id="toggler" size="sm">
      Help
    </Button>
    <UncontrolledCollapse toggler="#toggler">
      <div className="oi-help">{children}</div>
    </UncontrolledCollapse>
  </Row>
)
Help.propTypes = { children: PropTypes.element.isRequired }

export default Help
