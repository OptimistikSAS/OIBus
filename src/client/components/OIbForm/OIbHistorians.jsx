import React from 'react'
import PropTypes from 'prop-types'
import { Row, Col } from 'reactstrap'
import OIbTitle from './OIbTitle.jsx'
import Table from '../table/Table.jsx'

const OIbHistorians = ({ lastCompletedAt }) => {
  const renderTable = () => (
    <Table
      headers={['Type', 'Last Completed At', '']}
      rows={Object.keys(lastCompletedAt).map((key, i) => {
        const date = new Date(lastCompletedAt[key])
        return [
          {
            name: `key.${i}`,
            value: key,
          },
          {
            name: `value.${i}`,
            value: `${date.toLocaleDateString()} - ${date.toLocaleTimeString()} `,
          },
        ]
      })}
    />
  )

  return (
    <>
      <OIbTitle label="Historian support" />
      <Row>
        <Col md="6">
          {renderTable()}
        </Col>
      </Row>
    </>
  )
}

OIbHistorians.propTypes = { lastCompletedAt: PropTypes.object.isRequired }
OIbHistorians.defaultProps = {}

export default OIbHistorians
