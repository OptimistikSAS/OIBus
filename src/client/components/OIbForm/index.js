import OIbCheckBox from './OIbCheckBox.jsx'
import OIbInteger from './OIbInteger.jsx'
import OIbText from './OIbText.jsx'
import OIbTextArea from './OIbTextArea.jsx'
import OIbLink from './OIbLink.jsx'
import OIbPassword from './OIbPassword.jsx'
import OIbSelect from './OIbSelect.jsx'
import OIbScanMode from './OIbScanMode.jsx'
import OIbProxy from './OIbProxy.jsx'
import OIbTitle from './OIbTitle.jsx'
import OIbAuthentication from './OIbAuthentication.jsx'
import OIbTimezone from './OIbTimezone.jsx'
import OIbCron from './OIbCron.jsx'
import OIbLogLevel from './OIbLogLevel.jsx'

export {
  OIbCheckBox,
  OIbInteger,
  OIbText,
  OIbTextArea,
  OIbLink,
  OIbPassword,
  OIbSelect,
  OIbScanMode,
  OIbProxy,
  OIbTitle,
  OIbAuthentication,
  OIbTimezone,
  OIbCron,
  OIbLogLevel,
}
