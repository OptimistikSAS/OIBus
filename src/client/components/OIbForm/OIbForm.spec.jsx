import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import OIbForm from './OIbForm.jsx'

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('OIbForm', () => {
  test('check Form with value="{ form: {} }', () => {
    act(() => {
      ReactDOM.render(<OIbForm
        schema={{ form: {} }}
        values={{ object: 'value' }}
        onChange={() => (1)}
        name="name"
      />, container)
    })
    expect(container).toMatchSnapshot()
  })
})
