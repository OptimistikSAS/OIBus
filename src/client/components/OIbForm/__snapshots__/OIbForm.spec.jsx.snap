// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OIbForm check Form with value="{ form: {} } 1`] = `<div />`;
